Update colors, add the dark mode palette, and switch to CSS vars

zackkrida opened the pull request:
## Fixes
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, please consider opening one before creating this pull request. -->

Fixes #4325 
Fixes #4304
Fixes #4268

## Description
<!-- Concisely describe what the pull request does. -->
<!-- Add screenshots, videos, or other media to show the problem and the solution when appropriate. -->

This PR is the largest one for dark mode, in terms of the files touched. It replaces color name strings in nearly all of our Vue components. Specifically, it switches us from using explicit names like "pink" to semantic ones like "complementary". These semantic names have their underlying color automatically swapped out when dark mode is enabled.

This PR also makes some slight visual changes to our colors, most noticably by darkening our brand pink a tag. These updated colors were defined by @fcoveram in #4268.

The goals of this PR are as follows:

- Update to use CSS variables to define colors
- Update colors to the new hex values
- Update colors to use new names
- Update Vue components to use semantic names instead of explicit names 

As an ancillary benefit to this PR, you can view a fully functional dark mode by enabling the `FORCE_DARK_MODE` feature flag. This looks near-perfect, despite this not being a goal of this PR. It speaks to the power and quality of @fcoveram's color system that it already looks as good as it does.

As this is not a goal of the PR, this PR does not add new snapshots for the dark mode. We will test the dark mode on staging and locally, create new issue(s) to adjust the dark mode, and then "lock in" the appearance with a fresh set of dark mode snapshots added in #4305.

## Testing Instructions
<!-- Give steps for the reviewer to verify that this PR fixes the problem; or delete this section entirely. -->

- Review the changes to CSS variables in `frontend/src/styles/tailwind.css` and `frontend/tailwind.config.ts`
- Thoroughly click through the site in light mode. Review updated snapshots. Remember: Dark mode _can_ have issues that we will follow up on in a subsequent PR #4305.

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. -->

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.
- [x] I ran the DAG documentation generator (`ov just catalog/generate-docs` for catalog
      PRs) or the media properties generator (`ov just catalog/generate-docs media-props`
      for the catalog or `ov just api/generate-docs` for the API) where applicable.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

dhruvkb commented on frontend/src/components/VAudioThumbnail/VAudioThumbnail.vue:
```diff
@@ -1,7 +1,7 @@
   <!-- Should be wrapped by a fixed-width parent -->
   <div
-    class="relative h-0 w-full overflow-hidden bg-yellow pt-full"
+    class="relative h-0 w-full overflow-hidden bg-bg-fill-complementary pt-full"
```
I feel very weird about names like `text-text` and `bg-bg-fill-complementary`.  If some colours are specifically meant for specific usage, should we consider setting them in `theme.textColor` for text colours (and `theme.backgroundColor` for background colours) instead of the general `theme.colors`? We'll be able to avoid the `text-`/`bg-` prefix which leads to the repetition.

zackkrida replied:
@dhruvkb the textColor and backgroundColor keys solution is brilliant. I'll push that shortly. I've renamed the `bg-bg`, `text-text`, and so on to use `-base` instead.

dhruvkb reviewed (commented):
I'm still trying to get rid of a 500 error I'm experiencing locally so I need longer for a more thorough review, but the changes read OK to me and the snapshot diffs look good too.

Will we be adding snapshots for dark mode?

zackkrida commented:
@dhruvkb concerning dark mode snapshots, they'll be addressed in #4305, the rationale being, from the PR description:

> As an ancillary benefit to this PR, you can view a fully functional dark mode by enabling the FORCE_DARK_MODE feature flag. This looks near-perfect, despite this not being a goal of this PR. It speaks to the power and quality of @fcoveram's color system that it already looks as good as it does.
> 
> As this is not a goal of the PR, this PR does not add new snapshots for the dark mode. We will test the dark mode on staging and locally, create new issue(s) to adjust the dark mode, and then "lock in" the appearance with a fresh set of dark mode snapshots added in https://github.com/WordPress/openverse/issues/4305.

obulat reviewed (dismissed):
I've found several problems with colors on hover:

### The Clear filters button
#### Content settings modal
<img width="558" alt="Screenshot 2024-08-06 at 4 33 30 PM" src="https://github.com/user-attachments/assets/be29676b-8358-43b2-88a3-eca2fde38d2c">

#### Sidebar
<img width="350" alt="Screenshot 2024-08-06 at 4 43 57 PM" src="https://github.com/user-attachments/assets/748ef138-97dc-4ce9-9039-f5e6c6776e35">

### Items inside the modal
The external sources items should not be this dark on hover

<img width="533" alt="Screenshot 2024-08-06 at 4 35 30 PM" src="https://github.com/user-attachments/assets/ce9144b3-024e-4ab6-b9b4-453686ff14f4">

### The close buttons on dark background
#### Pages modal
<img width="557" alt="Screenshot 2024-08-06 at 4 32 14 PM" src="https://github.com/user-attachments/assets/fe958ec3-5d6f-436a-92e4-8df423174810">

#### Homepage banners
<img width="560" alt="Screenshot 2024-08-06 at 4 31 16 PM" src="https://github.com/user-attachments/assets/1d8dbbe4-0ec7-463c-a9b4-a081daac1b4d">

zackkrida commented:
@obulat thank you! 

I also just realized a huge component of this; our playwright vr component tests, which would likely find a lot of these hover and other component state issues, are disabled after Nuxt 3. 

I'll draft this and take some time to think about how to proceed there. Probably fixing the playwright component VR tests will need to happen prior to this PR.

zackkrida commented:
@dhruvkb and @obulat this is ready for re-review! I've eliminated the duplicate class name convention and fixed the button hover state identified by @obulat (it was one class name on one button variant impacting all of those screenshots).

dhruvkb reviewed (commented):
The changes look good to me, but I have a few suggestions. Since it works well, I can't justify blocking the PR.

zackkrida commented on frontend/src/styles/tailwind.css:
```diff
@@ -7,14 +7,218 @@
-  html {
-    @apply m-0 font-normal leading-loose text-dark-charcoal;
+  :root {
+    --black: #000;
+    --white: #fff;
```
I explicitly didn't want to do this to enforce not using them, as none of these opacity variants are used in the existing design.

zackkrida commented on frontend/tailwind.config.ts:
```diff
@@ -16,43 +22,147 @@ export default {
+      "wave-inactive": "var(--wave-inactive)",
+      "modal-layer": "var(--modal-layer)",
+
+      "info-1": "var(--info-1)",
+      "info-2": "var(--info-2)",
+      "info-3": "var(--info-3)",
+      "info-4": "var(--info-4)",
+      "info-5": "var(--info-5)",
+      "info-6": "var(--info-6)",
+      "info-7": "var(--info-7)",
+      "info-8": "var(--info-8)",
+      "info-9": "var(--info-9)",
+      "info-10": "var(--info-10)",
+      "info-11": "var(--info-11)",
+      "info-12": "var(--info-12)",
+      "info-13": "var(--info-13)",
```
Excellent, I'll do this 👍

zackkrida commented:
@dhruvkb I've adopted some of your suggestions:

- Removed `fill` from the background color names in the CSS variables _and_ the generated tailwind classes
- Prefixed color CSS vars with `color-`
- Switched to the [color object syntax](https://tailwindcss.com/docs/customizing-colors#color-object-syntax) for generated classes

Things I decided against:

- Storing RGB triplets instead of full color values. Our designs do not use the colors with varying opacity, so it doesn't make sense to me to produce these classes.
- Dynamically generating the color classes with JS. It's a fine change if someone wants to make it sometime, but I felt readability is actually simpler just to have a list of all the classes to go through.

dhruvkb reviewed (approved):
I agree with your rationale, so I won't hold this up any more. Thanks, looking forward to this everything dark mode!

zackkrida commented:
@obulat I've merged this. If you have any retroactive feedback do let me know. Thank you :smile:

fcoveram commented:
Thanks for reviewing this 🙏  I couldn't arrived timely.
